Opponent disconnected/End of game menus

caspernorrbin opened the pull request:
Added two menus, one for when the opponent disconnects and one for when the game is over. These are connected to their respective actions in the game. Currently, the `Play Again` option in the end menu does nothing. Functionality for this would need to be added in the future with with additional server logic.

`Menu` has also been extended to store the title text option to be able to change it after the fact, something that can be used to display dynamic menu titles, which is done in the end of game menu.

joelsiks commented on src/engine/UI/menu/Menu.cpp:
```diff
@@ -14,9 +14,16 @@ Menu::Menu(const std::string &menuTitle, const Color &foregroundColor,
     const auto titleOption = std::make_shared<TextOption>(
         TextOption(MENU_TITLE_NAME, Menu::commonDepthOrder, menuTitle));
 
+    m_titleOption = titleOption;
+
     this->AddDisplayable(titleOption);
```
I would combine this into 
```c++
m_titleOption = std::make_shared<TextOption>(
    TextOption(MENU_TITLE_NAME, Menu::commonDepthOrder, menuTitle));

this->AddDisplayable(m_titleOption);
```

joelsiks commented on src/engine/UI/menu/Menu.hpp:
```diff
@@ -19,6 +21,7 @@ class Menu : public IDisplayLayer {
 
 private:
     std::string m_menuTitle;
```
I would argue that we can remove this and replace with a getter that returns the text value of the title option instead.

joelsiks reviewed (changes requested):
Regarding closing the game using exit buttons, we also need to make sure we're handling the user clicking the "close window (X)" button correctly by releasing the static `GameData::engine` pointer.

Also, in relation to this, lets say we have two clients running the game, with one also running the server. If the non-server client closes the game, the correct menu is shown. However, if the client that is also running the sever is closed, we need a menu for the non-server client to say something along the lines of "Game server closed.." and an option to exit the game or go to the main menu.

Good job!

caspernorrbin commented:
In addition to the smaller fixes, I added a third "Game server closed" menu that displays on clients when the server connection disappears. I also added a third button "Return to menu" that switched back to the main menu. This needs more logic in the future to also reset the game state and not only switch the scene/menu

The engine pointer is now also released before exiting the program, allowing all destructors to run.

joelsiks reviewed (approved):
Nice work!
